core/src/main/java/io/questdb/cutlass/http/WaitProcessor.java

```diff
@@ -0,0 +1,221 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cutlass.http;
+
+import io.questdb.cairo.pool.ex.EntryUnavailableException;
+import io.questdb.mp.*;
+import io.questdb.std.time.MillisecondClock;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.PriorityQueue;
+import java.util.concurrent.locks.LockSupport;
+
+public class WaitProcessor implements RescheduleContext {
+
+    private static final int retryQueueLength = 4096;
+    private final RingQueue<RetryHolder> inQueue = new RingQueue<>(RetryHolder::new, retryQueueLength);
+    private final Sequence inPubSequence = new MPSequence(retryQueueLength);
+    private final Sequence inSubSequence = new SCSequence();
+    private final PriorityQueue<RetryHolder> nextRerun = new PriorityQueue<>(64);
+    private final RingQueue<RetryHolder> outQueue = new RingQueue<>(RetryHolder::new, retryQueueLength);
+    private final Sequence outPubSequence = new SPSequence(retryQueueLength);
+    private final Sequence outSubSequence = new MCSequence(retryQueueLength);
+    private final MillisecondClock clock;
+    private final long maxWaitCapMs;
+    private final double exponentialWaitMultiplier;
+
+    public WaitProcessor(JobRunner pool, WaitProcessorConfiguration configuration) {
+        this.clock = configuration.getClock();
+        this.maxWaitCapMs = configuration.getMaxWaitCapMs();
+        this.exponentialWaitMultiplier = configuration.getExponentialWaitMultiplier();
+
+        inPubSequence.then(inSubSequence).then(inPubSequence);
+        outPubSequence.then(outSubSequence).then(outPubSequence);
+
+        if (pool.getWorkerCount() > 0) {
+            int workerId = pool.getWorkerCount() - 1; // Last one lucky.
+            pool.assign(workerId, workerId1 -> processInQueue() || sendToOutQueue());
+        }
+    }
+
+    @Override
+    // This supposed to run in http execution thread / job
+    public void reschedule(Retry retry) {
+        reschedule(retry, 0, 0);
+    }
+
+    public void reschedule(Retry retry, int attempt, long waitStartMs) {
+        long now = clock.getTicks();
+        while (true) {
+            long cursor = inPubSequence.next();
+            // -2 = there was a contest for queue index and this thread has lost
+            if (cursor < -1) {
+                LockSupport.parkNanos(1);
+                continue;
+            }
+
+            // -1 = queue is empty. It means there are already too many retries waiting
+            // Send error to client.
+            if (cursor < 0) {
+                throw EntryUnavailableException.INSTANCE;
+            }
+
+            try {
+                RetryHolder retryHolder = inQueue.get(cursor);
+                retryHolder.retry = retry;
+                retryHolder.attempt = attempt;
```

Comment:
why don't we have these attributes on the context itself and not use holder? This way we don't really need to create objects for retries?

Reply:
Done